Formula/v8.rb

```diff
@@ -3,115 +3,77 @@
-
-    if build.with? "icu4c"
-      ENV.append "GYP_DEFINES", "use_system_icu=1"
-      i18nsupport = "i18nsupport=on"
+    (buildpath/"depot_tools").install buildpath.children - [buildpath/".brew_home"]
+    ENV["DEPOT_TOOLS_UPDATE"] = "0"
+    ENV.prepend_path "PATH", buildpath/"depot_tools"
+
+    system "gclient", "root"
+    system "gclient", "config", "--spec", <<-EOS.undent
+      solutions = [
+        {
+          "url": "https://chromium.googlesource.com/v8/v8.git",
+          "managed": False,
+          "name": "v8",
+          "deps_file": "DEPS",
+          "custom_deps": {},
+        },
+      ]
+      target_os = [ "mac" ]
+      target_os_only = True
+    EOS
+
+    cache_location = ENV["CACHE_DEPS"]
+    no_sync = ENV["NO_SYNC"]
+
+    if cache_location and File.directory?(cache_location)
+      system "cp", "-R", cache_location, "v8"
     else
-      i18nsupport = "i18nsupport=off"
+      system "gclient", "sync", "-vvv", "-j #{Hardware::CPU.cores}", "-r", "#{version}" unless no_sync
+
+      if cache_location and !File.directory?(cache_location)
+        system "cp", "-R", "v8", cache_location
+      end
     end
 
-    # fix up libv8.dylib install_name
-    # https://github.com/Homebrew/homebrew/issues/36571
-    # https://code.google.com/p/v8/issues/detail?id=3871
-    inreplace "tools/gyp/v8.gyp",
-              "'OTHER_LDFLAGS': ['-dynamiclib', '-all_load']",
-              "\\0, 'DYLIB_INSTALL_NAME_BASE': '#{opt_lib}'"
+    cd "v8" do
+      arch = MacOS.prefer_64_bit? ? "x64" : "x86"
+      output_name = "#{arch}.release"
+      output_path = "out.gn/#{output_name}"
 
-    (buildpath/"build/gyp").install resource("gyp")
-    (buildpath/"third_party/icu").install resource("icu")
-    (buildpath/"buildtools").install resource("buildtools")
-    (buildpath/"base/trace_event/common").install resource("common")
-    (buildpath/"tools/swarming_client").install resource("swarming_client")
-    (buildpath/"testing/gtest").install resource("gtest")
-    (buildpath/"testing/gmock").install resource("gmock")
-    (buildpath/"tools/clang").install resource("clang")
+      # Configure build
+      gn_args = {
+          is_debug: false,
+          is_component_build: true,
+          v8_use_external_startup_data: false
+      }
 
-    system "make", "native", "library=shared", "snapshot=on",
-                   "console=readline", i18nsupport,
-                   "strictaliasing=off"
+      system "gn gen #{output_path} --args=\"#{gn_args.map { |k, v| "#{k}=#{v}" }.join(' ')}\""
+      system "ninja", "-j #{Hardware::CPU.cores}", "-v", "-C", output_path, "d8"
 
-    include.install Dir["include/*"]
+      include.install Dir["include/*"]
 
-    cd "out/native" do
-      rm ["libgmock.a", "libgtest.a"]
-      lib.install Dir["lib*"]
-      bin.install "d8", "mksnapshot", "process", "shell" => "v8"
+      cd output_path do
+        lib.install Dir["lib*"]
+        bin.install "d8"
+      end
```

Comment:
@mathiasbynens pushed as separated commit. I can change it if will be needed :) Thanks for suggestion!